code/datums/elements/tool_renaming.dm

```diff
@@ -0,0 +1,67 @@
+/**
+ * Renaming tool element
+ *
+ * When using this tool on an object with UNIQUE_RENAME,
+ * lets the user rename/redesc it.
+ */
+/datum/element/tool_renaming
+
+/datum/element/tool_renaming/Attach(datum/target)
+	. = ..()
+	if(!isitem(target))
+		return ELEMENT_INCOMPATIBLE
+
+	RegisterSignal(target, COMSIG_ITEM_INTERACTING_WITH_ATOM, PROC_REF(attempt_rename))
+
+/datum/element/tool_renaming/Detach(datum/source)
+	. = ..()
+	UnregisterSignal(source, COMSIG_ITEM_INTERACTING_WITH_ATOM)
+
+/datum/element/tool_renaming/proc/attempt_rename(datum/source, mob/living/user, atom/interacting_with, list/modifiers)
+	SIGNAL_HANDLER
```

Comment:
Because this signal handler can't sleep, you're going to have to wrap everything involving input in a different proc called via INVOKE_ASYNC.

Reply:
Likely `ITEM_INTERACT_SUCCESS` to indicate a renaming attempt was completed, even if said attempt is cancelled

Reply:
agh, that's annoying, I'll look into it